misk-jobqueue-testing/src/main/kotlin/misk/jobqueue/FakeJobQueue.kt

```diff
@@ -150,14 +182,30 @@ class FakeJobQueue @Inject constructor(
       }
     }
 
+    // Reenqueue deadlettered jobs outside of the main loop to prevent an infinite loop.
     result.forEach { job ->
       if (job.deadLettered) {
-        deadletteredJobs.getOrPut(queueName, ::ConcurrentLinkedDeque).add(job)
+        deadletteredJobs.getOrPut(job.queueName, ::ConcurrentLinkedDeque).add(job)
       }
     }
-
     return result
   }
+
+  private fun pollNextJob(jobs: Queue<FakeJob>, considerDelays: Boolean): FakeJob? {
+    if (!considerDelays) {
+      return jobs.poll()
+    }
+    val now = clock.instant()!!
+    while (true) {
+      val job = jobs.firstOrNull {
+        it.deliveryDelay == null || !it.enqueuedAt.plus(it.deliveryDelay).isAfter(now)
+      }
+      // If remove() is false, then we lost race to another worker and should retry.
```

Comment:
I haven't find an atomic `find and remove`.

Reply:
Callers _probably_ shouldn’t be doing this concurrently?

Reply:
I think tests should avoid using concurrency at all, because it is not deterministic. But then  I'd replace ConcurrentHashMap by usual HashMap, etc. Otherwise using `ConcurrentHashMap` and then not caring about the rest is confusing. Can I remove all concurrency related stuff from this class? :)